doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,422 @@
+
+- Dump: Dumps the catalog contents in a human-readable format. Includes JSON for each key and value
+  that can be hand edited and then passed to the `edit` or `delete` commands.
+    - [optional arg] `target`: Write output to specified path. Default to stdout.
+- Edit: Edits a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+    - [arg] `value`: JSON encoded value of object.
+- Delete: Deletes a single catalog object.
+    - [arg] `object-type`: name of the object type to edit.
+    - [arg] `key`: JSON encoded key of object.
+- Upgrade Check: Checks if the specified catalog could be upgraded from its state to the catalog at
+  the version of this binary. Prints a success message or error message. Exits with 0 if the upgrade
+  would succeed, otherwise non-zero. Can be used on a running `environmentd`. Operates without
+  interfering with it or committing any data to that catalog.
+    - [optional arg] `cluster-replica-sizes`: Valid cluster replica sizes.
+
+The tool will have two modes:
+
+- Stash: Connects to a stash.
+- Persist: Connects to persist.
+
+The tool will initialize a `Box<dyn DurableCatalogState>` and use the trait methods to implement
+each functionality. `dyn DurableCatalogState` is safe because the trait is designed to be object
+safe.
+
+- Dump: Will use the `dump` method.
+- Edit and Delete: Will be implemented using a catalog transaction.
+- Upgrade Check: Will use the `check_open` method.
+
+NB: There is an open issue that you need two different versions of the stash debug tool to perform
+upgrade checks: https://github.com/MaterializeInc/materialize/issues/21121. This proposal will not
+fix that issue and suffers from the exact same issue.
+
+### Migrations
+
+This section is currently hand-wavy and needs the most additional design work.
+
+Stash migration ownership is currently tangled between the adapter crate and the stash crate,
+see [stash: Untangle the stash config collection](https://github.com/MaterializeInc/materialize/issues/20922).
+Ideally the adapter would be in charge of how and why to migrate the catalog contents while the
+stash provides the primitives needed to migrate a stash collection.
+
+Once the migrations are de-tangled, at a high level they will take the following steps:
+
+1. Use `impl From<objects_v(X - 1)::(Object)> for objects_v(X)::(Object)` to convert every object
+   from their old format into their new format.
+2. Use `mz_adapter::catalog::storage::Transaction` to perform any deletions, insertions, or updates
+   that require read/write semantics instead of `mz_stash::transaction::Transaction`.
+
+As a bonus, these migrations start to look very similar to the
+existing [catalog migrations](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs).
+We might be able to combine them into a single logical step instead of having two separate migration
```

Comment:
I thought about this in the past and I agree that it would be a good move. Maybe we add some sort of `async fn migrate_sql(...)` helper to the same Transaction type which implements the other migration helpers? `migrate_sql(...)` would basically be [`rewrite_items(...)`](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs#L30), but now you could call it in the context of of a `vX_to_vY` migration step. 

Just thinking out loud, there might be some complexity in migrating the serialization format at the same time as migrating the contents, but I think we could get around this by being clever with the type system, e.g. something like:

```
struct MigrationTransaction<Kind = Serialization> {
  inner: Transaction,
}

impl MigrationTransaction<Serialization> {
  async fn migrate_to(...) { ... }

  fn into_sql(self) -> MigrationTransaction<SQL> { ... }
}

impl MigrationTransaction<SQL> {
  async fn rewrite_items(...) { ... }
}
```

Once you turn your `MigrationTransaction` into a `MigrationTransaction<SQL>`, there would not be a way to turn it back into a `MigrationTransaction<Serialization>` so we can guarantee we only migrate SQL after we've already migrated the serialization format.